Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,483 @@
+            exponentSign = String(data[range]) == "-" ? -1 : 1
+        }
+
+        let decimalPartNSRange = match.range(at: REGroups.decimalPart.rawValue)
+        guard decimalPartNSRange.location != NSNotFound,
+            let decimalPartRange = Range(decimalPartNSRange, in: data) else {
+            throw BSONError.InvalidArgumentError(message: "Syntax Error: Missing digits in front of the exponent")
+        }
+        let decimalPart = String(data[decimalPartRange])
+        var digits = try Self.convertToDigitsArray(decimalPart)
+
+        var exponent = 0
+        let exponentPartRange = match.range(at: REGroups.exponentPart.rawValue)
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+            if exponent < Self.exponentMin {
+                exponent = Self.exponentMin
+            }
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.significandDigits {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(message: "Rounding Error: Cannot round exponent \(exponent) further")
+        }
+
+        guard digits.count <= Self.significandDigits else {
+            throw BSONError.InvalidArgumentError(
+                message: "Overflow Error: Value cannot exceed \(Self.significandDigits) digits"
+            )
+        }
+
+        var significandHiDigits = UInt64()
+        var significandLoDigits = UInt64()
+
+        let loDigits = Array(digits.suffix(Self.significandDigits / 2))
+        let hiDigits = Array(digits.dropLast(Self.significandDigits / 2))
+
+        if !loDigits.isEmpty {
+            significandLoDigits = UInt64(loDigits[0])
```

Comment:
might be worth extracting the logic that is used both here and for highDigits, e.g. into an `[UInt8]` extension method like `toUInt64()` or something